### PR TITLE
Correct handling of E_USER_ERROR as fatal error if registerErrorHandl…

### LINKED ISSUE #1669: Monolog doesn't handle E_USER_ERROR as fatal error if registerErrorHandler is called with callPrevious === false
Monolog version 2.6

Example:

```
$logger = new Monolog\Logger('main');
$logger->pushHandler(new Monolog\Handler\StreamHandler('php://stderr', Monolog\Logger::DEBUG));
$errorHandler = new ErrorHandler($logger);
$errorHandler->registerErrorHandler([], false); // false is for callPrevious
$errorHandler->registerExceptionHandler();

trigger_error('Error', E_USER_ERROR);
```

Description:

1. `$errorHandler->registerErrorHandler` sets `$this->previousErrorHandler` to `null`.
2. `ErrorHandler::handleError()` doesn't log error, but sets trace into the `$this->lastFatalTrace`. `handleError` assumes that the  `handleFatalError()` will be called.
3. `callPrevious` option prevents `ErrorHandler::handleError()` to start chaining, default PHP's error logger will not be called, `handleError` returns true. This prevents PHP from setting "last error" structure internally.
4. `ErrorHandler::handleFatalError()` catches E_USER_ERROR and tries to receive `$lastError = error_get_last();`, but there's no last error.
5. Both functions ignores error logging.

Scenario. I use official docker php image with error_log == 'php://stderr'. Also, I want to use Monolog with StreamHandler and write logs to stderr. Monolog logs errors twice with default settings (first time default PHP logger writes logs to stderr, second time Monolog does the same).

### PR DESCRIPTION
**Pull request by WoZ**
…er is called with callPrevious, fixes #1669

### CODE REVIEW

**Comment by WoZ** [comment]
It's impossible to test shutdown functions with PHPUnit, so I did a manual check.

**`callPrev == true`**
```php
use Monolog\ErrorHandler;

$logger = new Monolog\Logger('main');
$logger->pushHandler(new Monolog\Handler\StreamHandler('php://stderr', Monolog\Logger::DEBUG));

$errorHandler = new ErrorHandler($logger);
$errorHandler->registerErrorHandler([], true);
$errorHandler->registerFatalHandler();
$errorHandler->registerExceptionHandler();

trigger_error('Error', E_USER_ERROR);
```

Before Fix.
```
php-fpm_1  | NOTICE: PHP message: PHP Fatal error:  Error in /var/www/html/public/logger/index.php on line 15
php-fpm_1  | [2022-05-11T17:29:08.872957+00:00] main.ALERT: Fatal Error (E_USER_ERROR): Error {"code":256,"message":"Error","file":"/var/www/html/public/logger/index.php","line":15,"trace":[{"file":"/var/www/html/public/logger/index.php","line":15,"function":"trigger_error"}]} []
```

After Fix
```
php-fpm_1  | NOTICE: PHP message: PHP Fatal error:  Error in /var/www/html/public/logger/index.php on line 15
php-fpm_1  | [2022-05-11T17:25:43.907508+00:00] main.ALERT: Fatal Error (E_USER_ERROR): Error {"code":256,"message":"Error","file":"/var/www/html/public/logger/index.php","line":15,"trace":[{"file":"/var/www/html/public/logger/index.php","line":15,"function":"trigger_error"}]} []
```

**Conclusion: No changes, as expected** 

**`callPrev == false`**
```php
use Monolog\ErrorHandler;

$logger = new Monolog\Logger('main');
$logger->pushHandler(new Monolog\Handler\StreamHandler('php://stderr', Monolog\Logger::DEBUG));

$errorHandler = new ErrorHandler($logger);
$errorHandler->registerErrorHandler([], false);
$errorHandler->registerFatalHandler();
$errorHandler->registerExceptionHandler();

trigger_error('Error', E_USER_ERROR);
```

Before Fix.
Empty log.

After fix
```
php-fpm_1  | [2022-05-11T17:27:36.493950+00:00] main.ALERT: Fatal Error (E_USER_ERROR): Error {"code":256,"message":"Error","file":"/var/www/html/public/logger/index.php","line":15,"trace":[{"file":"/var/www/html/public/logger/index.php","line":15,"function":"trigger_error"}]} []
```

**Conclusion: Fixes the issue.** 

**Classical Fatal Error (non-cacthable by error handler**
```php
use Monolog\ErrorHandler;

$logger = new Monolog\Logger('main');
$logger->pushHandler(new Monolog\Handler\StreamHandler('php://stderr', Monolog\Logger::DEBUG));

$errorHandler = new ErrorHandler($logger);
$errorHandler->registerErrorHandler([], true);
$errorHandler->registerFatalHandler();
$errorHandler->registerExceptionHandler();

include('../error_handling/error_generators/fatal_error.php');
/** Code of error_handling/error_generators/fatal_error.php
 * <?php
 * function my_function(array $a) { echo $a; }
 * my_function(123.0);
 **/
```

Before fix.
```
php-fpm_1  | [2022-05-11T17:34:11.272915+00:00] main.ERROR: Uncaught Exception TypeError: "Argument 1 passed to my_function() must be of the type array, float given, called in /var/www/html/public/error_handling/error_generators/fatal_error.php on line 3" at /var/www/html/public/error_handling/error_generators/fatal_error.php line 2 {"exception":"[object] (TypeError(code: 0): Argument 1 passed to my_function() must be of the type array, float given, called in /var/www/html/public/error_handling/error_generators/fatal_error.php on line 3 at /var/www/html/public/error_handling/error_generators/fatal_error.php:2)"} []
```

After fix.
```
php-fpm_1  | [2022-05-11T17:36:50.263150+00:00] main.ERROR: Uncaught Exception TypeError: "Argument 1 passed to my_function() must be of the type array, float given, called in /var/www/html/public/error_handling/error_generators/fatal_error.php on line 3" at /var/www/html/public/error_handling/error_generators/fatal_error.php line 2 {"exception":"[object] (TypeError(code: 0): Argument 1 passed to my_function() must be of the type array, float given, called in /var/www/html/public/error_handling/error_generators/fatal_error.php on line 3 at /var/www/html/public/error_handling/error_generators/fatal_error.php:2)"} []
```

**Conclusion: Matching, No degradation**

**Comment by Seldaek** [comment]
Patch looks good aside from the type I commented on, thanks! 👍🏻

File: src/Monolog/ErrorHandler.php [hunk 1]
```diff
@@ -46,8 +46,8 @@ class ErrorHandler
     private $fatalLevel = LogLevel::ALERT;
     /** @var ?string */
     private $reservedMemory = null;
-    /** @var ?mixed */
-    private $lastFatalTrace;
+    /** @var ?array<mixed> */
```
[the remark this answers is not in the document]
Comment (WoZ) [reply src/Monolog/ErrorHandler.php]: As I see, you use PHPStan, not PSR-5. It's ok if so.

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**